### DISCUSSION TITLE
Websocket Routes overides uvicorns default Server header

### DISCUSSION THREAD

**Discussion by iudeen** [Kludex/uvicorn#1574] [Potential Issue] [answered]
This is a continuation of a discussion happened in FastAPI Discord Channel after an issue was reported by `Salaz#7667`.

When a Websocket route is created, there is a `Server` header that is present. Currently the server header has the python version. 

Sample: 
```text
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: SCtyBF9CdOAnFbass9dkRn6ZX0U=
Sec-WebSocket-Extensions: permessage-deflate
Date: Thu, 14 Jul 2022 07:15:48 GMT
Server: Python/3.10 websockets/10.3
```

On tracing it back, it was found that the websocket library adds this header. 

Interesting part is, for normal routes (http), the default headers added by Uvicorn looks like this:
```text
HTTP/1.1 200 OK
date: Thu, 14 Jul 2022 07:22:35 GMT
server: uvicorn
content-length: 4
content-type: application/json
```

This behavior exits even if `--no-server-header` is passed, only for Websockets route. 

One way I found to remove the Server header is to replace it by doing this:
```python
@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept(headers=[(b"Server", b"")])
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

### Test Scenario:

Webserver: uvicorn 0.18.2 with CPython 3.10.4 on Windows
Framework: FastAPI 0.78
OS: Windows
Websocket Client: Simple Web Socket Client

### Code Snippet
```python
from starlette.websockets import WebSocket
from fastapi import FastAPI, Form

@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

**Comment by iudeen** [comment]
IMHO, it should be considered a bug (minor) in Uvicorn simply for the fact that Uvicorn behaves differently for HTTP routes. 

I went through the code. Not exactly sure where we can set the appropriate server header before passing on to Web Sockets. 

As I mentioned in the question, setting the header (rather replacing header - as pointed by @Kludex in discord) during `websocket.accept` works. We just need to do it internally.

**Reply by iudeen** [reply]
To validate my argument, I did a test with other servers.

####  PGJones' [Hypercorn](https://pgjones.gitlab.io/hypercorn/) 

Here are the results.

```shell
>hypercorn app:app    
[2022-07-15 06:34:55 +0530] [1932] [INFO] Running on http://127.0.0.1:8000 (CTRL + C to quit)
```
```text
Request URL: ws://localhost:8000/ws
Request Method: GET
Status Code: 101 

HTTP/1.1 101
sec-websocket-extensions: permessage-deflate; client_max_window_bits=15
sec-websocket-accept: YlvIhcC5BulFuSpaElnogKYoBbs=
upgrade: WebSocket
connection: Upgrade
date: Fri, 15 Jul 2022 01:05:06 GMT
server: hypercorn-h11
```

####   [Daphne](https://github.com/django/daphne) 

Here are the results.

```shell
>daphne app:app     
2022-07-15 06:43:52,188 INFO     Starting server at tcp:port=8000:interface=127.0.0.1
2022-07-15 06:43:52,191 INFO     HTTP/2 support enabled
2022-07-15 06:43:52,191 INFO     Configuring endpoint tcp:port=8000:interface=127.0.0.1
2022-07-15 06:43:52,193 INFO     Listening on TCP address 127.0.0.1:8000
127.0.0.1:61903 - - [15/Jul/2022:06:44:07] "WSCONNECTING /ws" - -
127.0.0.1:61903 - - [15/Jul/2022:06:44:07] "WSCONNECT /ws" - -

```
```text
Request URL: ws://localhost:8000/ws
Request Method: GET
Status Code: 101 

HTTP/1.1 101 Switching Protocols
Server: Daphne
Upgrade: WebSocket
Connection: Upgrade
Sec-WebSocket-Accept: N8k+1GrZYeKkfMZxKgTfDTYDyAU=
```

If noticed, the server header set here is `hypercorn-h11` and `Daphne` respectively. 

I think, Uvicorn also has to do the same!

**Reply by iudeen** [reply]
I was able to get this fixed. Not sure if its the right way!

Check the PR in forked repo: https://github.com/iudeen/uvicorn/pull/1

If all good, I'll give a PR to main repo. (I tested with 0.18.2, modifying the code in my local)

**Reply by Kludex** [reply]
Hmm... There are two **potential** issues here:

1. Uvicorn still adds the server headers for websockets even when `--no-server-headers` is being used.
2. websockets use its own server headers.

I believe 1 is an issue, but I'm not sure we can disable them from uvicorn. An issue on websockets project may be worth it.

As for 2, I guess is also an issue, because the behavior should be the same on every protocol we use e.g. set "Server: uvicorn".

**Reply by iudeen** [reply]
@Kludex with the changes I made, the behavior is like this:

With `--no-server-header`:
```text
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: K03PH1V+r7YFwXw0z894QmNgTB0=
Sec-WebSocket-Extensions: permessage-deflate
date: Fri, 15 Jul 2022 05:58:19 GMT
Server: Python/3.10 websockets/10.3
```

Without `--no-server-header`:
```text
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: n0dTWast6MeO32dc/6ZAQuLxvzk=
Sec-WebSocket-Extensions: permessage-deflate
date: Fri, 15 Jul 2022 05:57:03 GMT
server: uvicorn
```

The Websockets library adds Server header if no header is present. Other ASGI Servers (Hypercorn, etc.) use Webproto instead of Websockets.

**Reply by iudeen** [reply]
Few ways I see we can handle is, 
1. if `--no-server-header` is passed Uvicorn should may be add `b""` for Websocket routes. (Not ideal, not elegant).
2. Recommend/add Websockets library to provision an API to remove the default server header.  Created an Issue: https://github.com/aaugustin/websockets/issues/1193
3. Move to webproto instead of websockets

**Reply by iudeen** [reply]
@Kludex, just curious, regardless of what is on the Websocket side, does the fix I made in the PR make sense and is proper?

**Reply by iudeen** [reply]
As the author mentioned in Websockets issue, I tried to add something like this in uvicorn in `uvicorn/config.py`

line#470
```python
        if isinstance(self.ws, str):
            if not self.server_header:
                import websockets.http
                websockets.http.USER_AGENT = ""
            ws_protocol_class = import_from_string(WS_PROTOCOLS[self.ws])
            self.ws_protocol_class: Optional[Type[asyncio.Protocol]] = ws_protocol_class
```

Now, with `--no-server-header`, we get 

```text
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: YeoIc7mbnQCz0JTWtmjyfnEJmgw=
Sec-WebSocket-Extensions: permessage-deflate
Server: 
date: Fri, 15 Jul 2022 09:06:38 GMT
```

**Reply by Kludex** [reply]
> @Kludex with the changes I made, the behavior is like this:
> 
> With `--no-server-header`:
> ```text
> HTTP/1.1 101 Switching Protocols
> Upgrade: websocket
> Connection: Upgrade
> Sec-WebSocket-Accept: K03PH1V+r7YFwXw0z894QmNgTB0=
> Sec-WebSocket-Extensions: permessage-deflate
> date: Fri, 15 Jul 2022 05:58:19 GMT
> Server: Python/3.10 websockets/10.3
> ```
> 
> Without `--no-server-header`:
> ```text
> HTTP/1.1 101 Switching Protocols
> Upgrade: websocket
> Connection: Upgrade
> Sec-WebSocket-Accept: n0dTWast6MeO32dc/6ZAQuLxvzk=
> Sec-WebSocket-Extensions: permessage-deflate
> date: Fri, 15 Jul 2022 05:57:03 GMT
> server: uvicorn
> ```
> 
> The Websockets library adds Server header if no header is present. Other ASGI Servers (Hypercorn, etc.) use Webproto instead of Websockets. 
> 

The way I see it, there's not much gain from doing this. The goal of the flag is not achieved. 😟

**Reply by iudeen** [reply]
:( agreed. Maximum we can get is an empty Server header.

**Reply by Kludex** [reply]
> Few ways I see we can handle is, 
> 1. if `--no-server-header` is passed Uvicorn should may be add `b""` for Websocket routes. (Not ideal, not elegant).
> 2. Recommend/add Websockets library to provision an API to remove the default server header.  Created an Issue: https://github.com/aaugustin/websockets/issues/1193
> 3. Move to webproto instead of websockets

1. It doesn't feel right to do that. Like, the flag is meant to remove the header completely.

2. Websockets maintainer set the label to enhancement on your issue. We can wait for it to be a thing and then come back to it later.

3. Uvicorn users can already use wsproto.

**Reply by Kludex** [reply]
I think we can consider this as a limitation for now, and even add a note about this on the documentation. Something like:

"On the websockets implementation, the server headers are not disabled by uvicorn."

**Reply by iudeen** [reply]
okay! let me add to the documentation then!

**Reply by iudeen** [reply]
Using
 ```shell
>uvicorn app:app --no-server-header --ws wsproto
``` 
```text
HTTP/1.1 101
Upgrade: WebSocket
Connection: Upgrade
Sec-WebSocket-Accept: EEHJutjs2IBv2mXrw45N10zYEZg=
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits=15
```

 ```shell
>uvicorn app:app  --ws wsproto
``` 
```
HTTP/1.1 101
Upgrade: WebSocket
Connection: Upgrade
Sec-WebSocket-Accept: I2DZEBqAyk74ZXuTII7qLhd+LG0=
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits=15
```

No `Server` header in both cases!

**Reply by Kludex** [reply]
Maybe we need to see if not adding the server and date headers on websockets was on purpose... 🤔

**Reply by iudeen** [reply]
Uhmm! this is getting interesting! How do we confirm this?

**Reply by Kludex** [reply]
Checking the discussion on the PR that introduced the server header.

**Reply by iudeen** [reply]
I don't think anything is mentioned about it explicitly here: https://github.com/encode/uvicorn/pull/818

**Reply by Kludex** [reply]
That's the PR that added the possibility to not add the headers. What about the PR that introduced the server headers?

**Reply by iudeen** [reply]
Not able to find :/

**Reply by iudeen** [reply]
I searched with keywords `header`, `server_header` and even `default_header`. May be I overlooked 😔

**Reply by iudeen** [reply]
I went through https://github.com/encode/uvicorn/commit/07c6c18945ab4f4167118e717e197adce3aee3da, I think headers were left out. Not exactly sure why..

**Reply by iudeen** [reply]
I have started a PR for this. ~~https://github.com/encode/uvicorn/pull/1581~~ https://github.com/encode/uvicorn/pull/1606

I'd need some insights from you @Kludex if this makes sense or it'd be ideal to stall the PR till we have a clear idea on this implementation.

**Reply by iudeen** [reply]
> I went through https://github.com/encode/uvicorn/commit/07c6c18945ab4f4167118e717e197adce3aee3da, I think headers were left out. Not exactly sure why..

I this the original commit left out headers because `Websockets` did not provide an API for it. Just my wild guess. Any thoughts? @Kludex ?

**Comment by iudeen** [comment, marked as answer]
This is fixed in https://github.com/encode/uvicorn/pull/1606
